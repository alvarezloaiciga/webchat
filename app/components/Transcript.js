import React, {Component} from 'react';
import Message from 'Message/Message';
import PlatformEvent from './PlatformEvent';
import quiqOptions, {getMessage} from 'Common/QuiqOptions';
import {connect} from 'react-redux';
import findLastIndex from 'lodash/findLastIndex';
import {getTranscript, getPlatformEvents, getLatestConversationIsSpam} from 'reducers/chat';
import {messageTypes} from 'Common/Constants';
import {setInputtingEmail} from 'actions/chatActions';
import type {Message as MessageType, ChatState, Event, ChatConfiguration} from 'Common/types';
import './styles/Transcript.scss';

export type TranscriptProps = {
  /*eslint-disable react/no-unused-prop-types*/
  transcript: Array<MessageType>,
  platformEvents: Array<Event>,
  latestConversationIsSpam: boolean,
  agentTyping: boolean,
  configuration: ChatConfiguration,
  setInputtingEmail: (inputtingEmail: boolean) => void,
  /*eslint-enable react/no-unused-prop-types*/
};

export type TranscriptState = {
  inputtingEmail: boolean,
};

export class Transcript extends Component {
  props: TranscriptProps;
  state: TranscriptState = {
    inputtingEmail: false,
  };
  scrollLock: boolean = false;
  transcript: HTMLElement;

  componentDidMount() {
    this.scrollToBottom();

    // Listen for scroll, set scrollLock flag
    if (this.transcript) {
      this.transcript.addEventListener(
        'wheel',
        () => {
          this.scrollLock = true;
        },
        {passive: true},
      );
    }
  }

  scrollToBottom = () => {
    if (!this.transcript || this.scrollLock) return;

    this.transcript.scrollTop = this.transcript.scrollHeight;
  };

  componentDidUpdate(prevProps) {
    // Scroll to the bottom if you get a new message or new visible platform event
    const oldCount = this.getConversationElementsForDisplay(prevProps).length;
    const newCount = this.getConversationElementsForDisplay().length;
    if (newCount > oldCount) {
      this.scrollLock = false;
      this.scrollToBottom();
    }
  }

  getConversationElementsForDisplay = (
    props: TranscriptProps = this.props,
  ): Array<Message | Event> =>
    [...props.transcript, ...props.platformEvents].filter(e =>
      ['Attachment', 'Text', 'SendTranscript', 'End', 'Spam'].includes(e.type),
    );

  render() {
    const {colors} = quiqOptions;

    const messagesAndEvents = this.getConversationElementsForDisplay().sort(
      (a, b) => a.timestamp - b.timestamp,
    );

    const agentHasResponded = this.props.transcript.some(m => m.authorType === 'User');

    const lastEndEventIdx = findLastIndex(messagesAndEvents, e => e.type === 'End');

    const lastMessageIndex = findLastIndex(messagesAndEvents, e =>
      ['Attachment', 'Text'].includes(e.type),
    );

    const displayElements = messagesAndEvents.map((a, idx) => {
      if (a.type === 'Attachment' || a.type === 'Text') {
        return (
          <Message key={a.localKey || a.id} message={a} scrollToBottom={this.scrollToBottom} />
        );
      }

      // If this is last End event and there are no messages afterwards, show the inline email button
      if (
        idx === lastEndEventIdx && // Is this event the last end event?
        lastMessageIndex < lastEndEventIdx && // Must not be any messages after end event
        agentHasResponded && // Agent must have responded at some point ion the entire transcript
        !this.props.latestConversationIsSpam
      ) {
        // Current conversation must not be spam
        return (
          <PlatformEvent
            event={a}
            key={a.id}
            action={() => this.props.setInputtingEmail(true)}
            actionLabel={getMessage(messageTypes.emailTranscriptInlineButton)}
          />
        );
      }

      return <PlatformEvent event={a} key={a.id} />;
    });

    // Append an AgentTyping message if needed
    if (this.props.agentTyping) {
      displayElements.push(
        <Message
          key="agentTyping"
          message={{authorType: 'Agent', type: 'AgentTyping'}}
          scrollToBottom={this.scrollToBottom}
        />,
      );
    }

    return (
      <div
        className="Transcript"
        style={{
          backgroundColor: colors.transcriptBackground,
          paddingBottom: messagesAndEvents.length > 0 ? '10px' : '0px',
        }}
        ref={n => {
          this.transcript = n;
        }}
      >
        {displayElements}
      </div>
    );
  }
}

const mapDispatchToProps = {
  setInputtingEmail,
};

export default connect(
  (state: ChatState) => ({
    latestConversationIsSpam: getLatestConversationIsSpam(state),
    transcript: getTranscript(state),
    agentTyping: state.agentTyping,
    platformEvents: getPlatformEvents(state),
    configuration: state.configuration,
  }),
  mapDispatchToProps,
)(Transcript);
