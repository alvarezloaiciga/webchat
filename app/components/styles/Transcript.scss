.Transcript {
  display: flex;
  flex-direction: column;
  overflow-y: auto;
  flex: 1 1 auto;
}

// Total Width: 400px
// Total Height: 600px
// MessageForm: 120px
// Banner: 40px
// Transcript: 440px
.noFlexbox .Transcript {
  margin-top: 40px;
  height: 440px;
  min-height: 440px;
  max-height: 440px;
}
