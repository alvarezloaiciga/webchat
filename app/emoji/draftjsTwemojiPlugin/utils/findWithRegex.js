const findWithRegex = (regex, contentBlock, callback) => {
  // Get the text from the contentBlock
  const text = contentBlock.getText();
  let matchArr;
  let start; // eslint-disable-line
  // Go through all matches in the text and return the indices to the callback
  // eslint-disable-next-line no-cond-assign
  while ((matchArr = regex.exec(text)) !== null) {
    start = matchArr.index;
    callback(start, start + matchArr[0].length);
  }
};

export default findWithRegex;
