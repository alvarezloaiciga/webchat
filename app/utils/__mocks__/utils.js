export const getBrowserName = () => 'Chrome';
export const getMajor = () => 60;
export const getDeviceType = () => 'PC';
export const getOSName = () => 'Windows';
export const getUAInfo = () => ({
  browser: {
    name: 'Chrome',
    major: 60,
  },
  os: {
    name: 'Mac OS',
    version: 10,
  },
  device: {
    model: 'iPad',
    vendor: 'Apple',
    type: 'tablet',
  },
  cpu: {
    architecture: 'amd64',
  },
  engine: {
    name: 'WebKit',
    version: 537.36,
  },
  navigator:
    'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.81 Safari/537.36',
});
export const compatibilityMode = () => false;
export const isIE9 = () => false;
export const isIE10 = () => false;
export const nonCompatibleBrowser = () => false;
export const supportsFlexbox = () => true;
export const supportsSVG = () => true;
export const displayError = (error: string) => error;
export const isIEorSafari = () => false;
