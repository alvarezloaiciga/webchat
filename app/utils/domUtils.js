import ReactDOM, {render} from 'react-dom';
import React from 'react';
import Routes from 'Routes';
import {quiqContainerId} from 'Common/Constants';
import {AppContainer} from 'react-hot-loader';
import {IntlProvider} from 'react-intl';
import Redbox from 'redbox-react';
import {Provider} from 'react-redux';

export const appIsMounted = (): boolean => {
  const container = document.getElementById(quiqContainerId);
  if (!container) return false;
  return container.childElementCount > 0;
};

export const destructApp = () => {
  ReactDOM.unmountComponentAtNode(document.getElementById(quiqContainerId));
};

export const constructApp = store => {
  render(
    <Provider store={store}>
      <IntlProvider locale="en">
        <AppContainer errorReporter={Redbox}>
          <Routes />
        </AppContainer>
      </IntlProvider>
    </Provider>,
    document.getElementById(quiqContainerId),
  );

  if (module.hot) {
    module.hot.accept('Routes', () => {
      const NextApp = require('Routes').default; // eslint-disable-line global-require

      render(
        <Provider store={store}>
          <IntlProvider locale="en">
            <AppContainer errorReporter={Redbox}>
              <NextApp />
            </AppContainer>
          </IntlProvider>
        </Provider>,
        document.getElementById(quiqContainerId),
      );
    });
  }
};

export const removeElement = (e: HTMLElement) => {
  e.parentNode.removeChild(e);
};

export const reloadApp = store => {
  destructApp();
  constructApp(store);
};

export const setFavicon = (iconUrl: string, appleTouchUrl: ?string) => {
  // Remove all icons
  const icons = [
    ...Array.from(document.querySelectorAll("link[rel='icon']")),
    ...Array.from(document.querySelectorAll("link[rel='apple-touch-icon']")),
  ];
  icons.forEach(icon => {
    removeElement(icon);
  });

  // Build icon
  const icon = document.createElement('link');
  icon.rel = 'icon';
  icon.href = iconUrl;

  // Build apple touch icon
  const appleTouchIcon = document.createElement('link');
  appleTouchIcon.rel = 'apple-touch-icon';
  appleTouchIcon.href = appleTouchUrl || iconUrl;

  // NOTE HARD: The apple-touch-icon must come BEFORE the regular icon.
  // Chrome looks for the last meta tag with a rel containing "icon" to try and use as title bar icon.
  const head = document.getElementsByTagName('head')[0];
  head.appendChild(appleTouchIcon);
  head.appendChild(icon);
};

export const setApplicationName = (name: string) => {
  const appName = document.querySelector("meta[name='application-name']");
  if (appName) {
    removeElement(appName);
  }
  const newAppName = document.createElement('meta');
  newAppName.name = 'application-name';
  newAppName.content = name;
  document.getElementsByTagName('head')[0].appendChild(newAppName);
};

export const setBrowserThemeColor = (color: string) => {
  const colorTag = document.querySelector("meta[name='theme-color']");
  if (colorTag) {
    removeElement(colorTag);
  }
  const newColor = document.createElement('meta');
  newColor.name = 'theme-color';
  newColor.content = color;
  document.getElementsByTagName('head')[0].appendChild(newColor);
};
