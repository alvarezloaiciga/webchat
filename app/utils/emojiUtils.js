import {emojiIndex} from 'emoji-mart';
import emojiRegexFactory from 'emoji-regex';
import invertBy from 'lodash/invertBy';
import type {EmojiMetadata, Emoji} from 'Common/Types';
import quiqOptions from 'Common/QuiqOptions';

const emojiRegex = emojiRegexFactory();

// Build lookup table of {unifiedCode : emojiId}
const emojiIdByUnified = {};
Object.keys(emojiIndex.emojis).forEach(id => {
  const unified = emojiIndex.emojis[id].unified.toLowerCase();
  emojiIdByUnified[unified] = id;
});

export const isSingleEmoji = (text: string): boolean => {
  // We know a string contains only one emoji iff
  // 1) An emoji regex finds only one match in the string, and
  // 2) The number of characters in that emoji === the number of characters in the string
  const emojis = text.match(emojiRegex);

  if (!emojis)
    return false;

  const emojiCharCount = emojis.reduce((count, e) => count + e.length, 0);
  return emojis.length === 1 && text.length === emojiCharCount;
};

export const convertUnicodeToEmojiObject = (u: string): ?Emoji => {
  const emojiId = Object.keys(emojiIndex.emojis).find(k => emojiIndex.emojis[k].native === u);
  return emojiIndex.emojis[emojiId];
};

// This function is used for filtering emojis in our picker, as well as redacting emojis prior to sending a message.
export const emojiFilter = (e: EmojiMetadata | Emoji | string) => {
  let emojiId;
  // emoji-mart will call this function when user is searching, but it only passes unicode string:
  if (typeof e === 'string') {
    emojiId =  emojiIdByUnified[e.toLowerCase()];
  } else {
    // NOTE: In emoji-mart land, id === short_codes[0]
    emojiId = e.id || e.short_names[0];
  }

  // Don't include any emojis that don't have id or shortcode (we don't know what they are)
  if (!emojiId)
    return false;

  const {includeEmojis, excludeEmojis} = quiqOptions;
  // Prioritize include over exclude. Only one of the should be defined by customer, however.
  if (Array.isArray(includeEmojis)) {
    return includeEmojis.includes(emojiId);
  }
  else if (Array.isArray(excludeEmojis)) {
    return !excludeEmojis.includes(emojiId);
  }

  // If neither exclude nor include was defined, always include emoji.
  return true;
};
