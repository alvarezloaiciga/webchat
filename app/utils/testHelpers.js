// @flow
import type {Message, IntlObject, IntlMessage} from 'types';

export const getMockMessage = (
  // eslint-disable-line import/prefer-default-export
  id?: number = 0,
  overrides?: Message | {} = {},
) => {
  const message = {
    authorType: id % 2 === 0 ? 'Customer' : 'Agent',
    text: `text--${id}`,
    id: `id--${id}`,
    timestamp: id,
    type: 'Text',
  };

  return Object.assign(message, overrides);
};

/**
 * Stub out formatMessage so that it will return the defaultMessage with basic variable replacement
 */
const formatTestMessage = (msg: IntlMessage, values?: {[key: string]: string}) => {
  if (!values) {
    return msg.defaultMessage;
  }

  let returnValue = msg.defaultMessage;
  Object.keys(values).forEach(key => {
    if (values) {
      const value = values[key];
      returnValue = returnValue.replace(`{${key}}`, value);
    }
  });

  return returnValue;
};

export const TestIntlObject: IntlObject = {
  formatMessage: formatTestMessage,
  formatRelative: value => value.toString(),
  formatTime: timestamp => timestamp.toString(),
  formatDate: timestamp => timestamp.toString(),
};
