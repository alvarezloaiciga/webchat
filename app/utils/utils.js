// @flow
declare var Modernizr: Object;
import {UAParser} from 'ua-parser-js';
import type {BrowserNames, DeviceTypes, OSNames} from 'types';

const parser = new UAParser();

export const getBrowserName = (): BrowserNames => parser.getResult().browser.name;

export const getMajor = (): number => parseInt(parser.getResult().browser.major, 10);

export const getDeviceType = (): DeviceTypes => parser.getResult().device.type;

export const getOSName = (): OSNames => parser.getResult().os.name;

export const getUAInfo = () => parser.getResult();

export const compatibilityMode = () => {
  const compatList = [
    {
      name: 'IE',
      major: 10,
    },
  ];
  const {name, major} = parser.getResult().browser;
  return !!compatList.find(i => i.name === name && parseInt(major, 10) <= i.major);
};

export const isIE9 = () => getBrowserName() === 'IE' && getMajor() <= 9;
export const isIE10 = () => getBrowserName() === 'IE' && getMajor() === 10;

export const nonCompatibleBrowser = () => getBrowserName() === 'IE' && getMajor() < 9;
// Modernizr says IE10 doesn't support flexbox.
// It kind of does, at least for what we need it for... so go ahead and ignore Modernizr in that case
export const supportsFlexbox = () => isIE10() || Modernizr.flexbox;
