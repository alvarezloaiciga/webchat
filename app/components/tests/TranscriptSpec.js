// @flow
import React from 'react';
import { getMockMessage } from 'utils/testHelpers';
import type { TranscriptProps } from '../Transcript';
import Transcript from '../Transcript';
import { shallow } from 'enzyme';
import type { ShallowWrapper } from 'enzyme';


describe('Transcript component', () => {
  let wrapper:ShallowWrapper;
  let testProps:TranscriptProps;
  let render: () => void;

  beforeEach(() => {
    testProps = {
      messages: [getMockMessage(1), getMockMessage(2)],
      agentTyping: true,
    };
    render = () => {
      wrapper = shallow(<Transcript {...testProps}/>);
    };
  });

  describe('rendering', () => {
    beforeEach(() => {
      render();
    });

    it('renders with default props', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
