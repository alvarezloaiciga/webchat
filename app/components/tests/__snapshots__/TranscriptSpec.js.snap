// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transcript component rendering renders with default props 1`] = `
<div
  className="Transcript"
>
  <Message
    message={
      Object {
        "authorType": "Agent",
        "body": "body--1",
        "id": "id--1",
        "timestamp": 1,
        "type": "Text",
      }
    }
  />
  <Message
    message={
      Object {
        "authorType": "Guest",
        "body": "body--2",
        "id": "id--2",
        "timestamp": 2,
        "type": "Text",
      }
    }
  />
</div>
`;
